src/mcts/node.h

```diff
@@ -570,6 +553,108 @@ class Edge_Iterator : public EdgeAndNode {
   uint16_t total_count_ = 0;
 };
 
+// TODO(crem) Replace this with less hacky iterator once we support C++17.
+// This class has multiple hypostases within one class:
+// * Range (begin() and end() functions)
+// * Iterator (operator++() and operator*())
+// It's more customary to have those as two classes, but
+// creating zoo of classes and copying them around while iterating seems
+// excessive.
+//
+// All functions are not thread safe (must be externally synchronized).
+template <bool is_const>
+class Node_Iterator {
+ public:
+  // Creates "end()" iterator.
+  Node_Iterator() {}
+
+  // Creates "begin()" iterator. Also happens to be a range constructor.
+  // child_ptr will be nullptr if parent_node is solid children.
+  Node_Iterator(const Node& parent_node, Node* child_ptr)
+      : node_ptr_(child_ptr),
+        total_count_(parent_node.num_edges_),
+        solid_(parent_node.solid_children_) {
+    if (node_ptr_ != nullptr && node_ptr_->GetN() == 0) {
+      operator++();
+    }
+  }
+  // These are technically wrong, but are usable to compare with end().
+  bool operator==(const Node_Iterator<is_const>& other) const {
+    return node_ptr_ == other.node_ptr_;
+  }
+  bool operator!=(const Node_Iterator<is_const>& other) const {
+    return node_ptr_ != other.node_ptr_;
+  }
+
+  // Function to support range interface.
+  Node_Iterator<is_const> begin() { return *this; }
+  Node_Iterator<is_const> end() { return {}; }
+
+  // Functions to support iterator interface.
+  // Equality comparison operators are inherited from EdgeAndNode.
+  void operator++() {
+    if (solid_) {
+      while (++current_idx_ != total_count_ &&
+             node_ptr_[current_idx_].GetN() == 0) {
+        if (node_ptr_[current_idx_].GetNInFlight() == 0) {
+          // Once there is not even n in flight, we can skip to the end.
+          current_idx_ = total_count_;
+          break;
+        }
+      }
+      if (current_idx_ == total_count_) {
+        node_ptr_ = nullptr;
+      }
+    } else {
+      do {
+        node_ptr_ = node_ptr_->sibling_.get();
+        // If n started is 0, can jump direct to end.
```

Comment:
Is this a request to improve documentation or do you not know about the sorted edges optimization?
(Since edges are sorted by P, the edges can be partitioned into two sections NStarted > 0, and NStarted == 0) Anytime we need a new edge to start its the first from the NStarted == 0 section, since highest P wins.

Reply:
I added half a paragraph :P - but done in 1490.